### PR TITLE
texlive: add licensing information

### PR DESCRIPTION
**Pull request by apfelkuchen6**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by veprbl** [comment]
Do we really need to aggregate this information? My understanding is: it is guaranteed that texlive upstream selects subset of CTAN packages with free licenses.

File: pkgs/tools/typesetting/tex/texlive/combine.nix [hunk 1]
```diff
@@ -17,6 +17,14 @@ let
   pkgList = rec {
     combined = combinePkgs (lib.attrValues pkgSet);
     all = lib.filter pkgFilter combined;
+    licenses = lib.foldl (acc: pkg:
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/combine.nix:20]: Maybe `mergeLicenses`, or `concatLicenses`? I find the double fold hard to read, but as long as the performance is reasonable, it seems ok.

File: pkgs/tools/typesetting/tex/texlive/tl2nix.sed [hunk 2]
```diff
@@ -36,6 +36,35 @@ $a}
...
+  /^catalogue-license/{
+        # wrap licenses in quotes
+        s/ ([^ ]+)/ "\1"/g
+        # adjust naming as in nixpkgs, the full text of the licenses are available at https://www.ctan.org/license/${licenseName}
+        s/"(cc-by(-sa)?-[1-4])"/"\10"/g
+        s/"apache2"/"asl20"/g
+        s/"artistic"/"artistic1"/g
+        s/"bsd"/"bsd3"/g          # license text does not match exactly, but is pretty close
+        s/"bsd4"/"bsdOriginal"/g
+        s/"collection"//g         # used for the empty collection metapackages
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/tl2nix.sed:48]: Actually collection refers to packages made up of disparate sources with different licenses (e.g. https://ctan.org/pkg/eijkhout). The collection meta packages do not have licenses attached. I am not sure, but maybe we can use `"free"` here. We can assume that if the package is on TeX Live, then is is at least `"free"` – can we?

File: pkgs/tools/typesetting/tex/texlive/tl2nix.sed [hunk 3]
```diff
@@ -36,6 +36,35 @@ $a}
   /^catalogue-version/s/[\#,:\(\)]//g
   s/^catalogue-version_(.*)/  version = "\1";/p
 
+  /^catalogue-license/{
+        # wrap licenses in quotes
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/tl2nix.sed:40]: The rest of the script uses 2 spaces for indentation.

**Review by xworld21 — CHANGES_REQUESTED** [review]
Awesome. This was way overdue. I double checked the licenses and flagged a few dodgy ones.

File: pkgs/tools/typesetting/tex/texlive/tl2nix.sed [hunk 4]
```diff
@@ -36,6 +36,35 @@ $a}
...
+        s/"collection"//g         # used for the empty collection metapackages
+        s/"fdl"/"fdl13Only"/g
+        s/"gfl"/"lppl13c"/g       # https://www.ctan.org/license/gfl
+        s/"gfsl"/"lppl13a"/g      # https://www.ctan.org/license/gfsl
+        s/"gpl1?"/"gpl1Only"/g
+        s/"gpl2\+"/"gpl2Plus"/g
+        s/"gpl3\+"/"gpl3Plus"/g
+        s/"lgpl"/"lgpl2"/g
+        s/"lgpl2\.1"/"lgpl21"/g
+        s/"lppl"/"lppl1"/g
```
[the remark this answers is not in the document]
Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/tl2nix.sed:57]: Mh, [Fedora](https://packages.fedoraproject.org/pkgs/texlive/texlive-12many/) uses lppl-1.3c for 12many. Picking the newest version if the version isn't specified is probably the most reasonable approach. Similarly, the lppl website (but not the license text) says „If a work refers to LPPL 1.3 as its license (like the example above) this is interpreted as the latest version of the 1.3 license, i.e., version 1.3c“. The differences between the versions also appear to be minimal.

File: pkgs/tools/typesetting/tex/texlive/tl2nix.sed (hunk 2 above)
Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/tl2nix.sed:48]: Oh, I just looked at the license description and assumed that it was used for the collections in texlive. Good catch!
TeXLive's [LICENSE.TL](https://tug.org/texlive/LICENSE.TL) file says that to their knowledge, everything in texlive is free software. I'll go ahead and use "free" for this then

**Review by drupol — APPROVED** [review]

**Comment by vcunat** [comment]
At least the other parts of this PR don't seem to have a noticeable impact on evaluation costs:
https://hydra.nixos.org/job/nixpkgs/trunk/metrics#tabs-charts
(the charts only shows evaluations that succeeded, of course)

**Comment by xworld21** [comment]
> I reverted the texlive.combine change, at least for now. Big jobsets were unable to evaluate on Hydra, crashing with over 40k C++ frames on the stack. The computation was probably way too expensive.

Ah, so that's why I was getting segfaults yesterday (on Darwin). @apfelkuchen6 you need to cut down the recursion – maybe with `foldl'`, or at worst `genericClosure`.

**Comment by wentasah** [comment]
This causes `attribute 'noinfo' missing` error when building with *doc* packages.

It can be reproduced with:
```sh
nix-instantiate --expr 'with import <nixpkgs> {}; texlive.combine { inherit (texlive) scheme-full; pkgFilter = (pkg: builtins.elem pkg.tlType [ "run" "bin" "doc"]); }'
```

This command results in the following error message:
```
error:
       … while calling the 'derivationStrict' builtin

         at //builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'texlive-combined-2022'
         whose name attribute is located at /nix/store/6y81y7ishz9az02ix5a815s9dglap7lx-nixpkgs/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:303:7

       … while evaluating attribute 'passAsFile' of derivation 'texlive-combined-2022'

         at /nix/store/6y81y7ishz9az02ix5a815s9dglap7lx-nixpkgs/nixpkgs/pkgs/build-support/trivial-builders/default.nix:88:7:

           87|       inherit buildCommand name;
           88|       passAsFile = [ "buildCommand" ]
             |       ^
           89|         ++ (derivationArgs.passAsFile or []);

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'noinfo' missing

       at /nix/store/6y81y7ishz9az02ix5a815s9dglap7lx-nixpkgs/nixpkgs/pkgs/tools/typesetting/tex/texlive/default.nix:188:31:

          187|           meta = {
          188|             license = map (x: lib.licenses.${x}) (args.license or []);
             |                               ^
          189|           };
```

**Pull request merged**